test/uk/gov/hmrc/tai/model/domain/TaxCodeIncomeSpec.scala

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright 2019 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.tai.model.domain
+
+import org.joda.time.LocalDate
+import org.scalatestplus.play.PlaySpec
+import play.api.libs.json.{JsObject, Json}
+import uk.gov.hmrc.tai.factory.TaxCodeIncomeFactory
+import uk.gov.hmrc.tai.model.domain.income.OtherBasisOperation
+
+class TaxCodeIncomeSpec extends PlaySpec {
+
+  val taxCodeIncome = TaxCodeIncomeFactory.create
+
+  "TaxCodeIncomeSource taxCodeWithEmergencySuffix" must {
+    "return the taxCode WITH X suffix" when {
+      "the basis operation is week1Month1" in {
+        taxCodeIncome.taxCodeWithEmergencySuffix mustBe "K100X"
+      }
+    }
+    "return the taxCode WITHOUT X suffix" when {
+      "the basis operation is NOT week1Month1" in {
+        val model = taxCodeIncome.copy(basisOperation = OtherBasisOperation)
+        model.taxCodeWithEmergencySuffix mustBe "K100"
+      }
+    }
+  }
+
+  "TaxCodeIncome Writes" must {
+    "write the taxCode correctly" when {
```

Comment:
Could you write a `there and back again` test to ensure that we don't change data during the serialization and deserialization process (except for the tax code). Some thing like 
```
"ensure no unexpected data is changed during serialization and deserialization" in {
  val data = TaxCodeIncomeFactory.create
  Json.toJson(data).as[TaxCodeIncome].copy(taxCode = data.taxCode) mustEqual data
}
```

Reply:
Looking at this today and we think this extra test is probably unnecessary, as was our previous implementation of an implicit reads on TaxCodeIncome. This is because there is already a reads defined for TaxCodeIncome in TaxCodeIncomeHodFormatters:

`  val taxCodeIncomeSourceReads = new Reads[TaxCodeIncome] {
    override def reads(json: JsValue) = {
      val incomeSourceType = taxCodeIncomeType(json)
      val employmentId = (json \ "employmentId").asOpt[Int]
      val amount = totalTaxableIncome(json, employmentId).getOrElse(BigDecimal(0))
      val description = incomeSourceType.toString
      val taxCode = (json \ "taxCode").asOpt[String].getOrElse("")
      val name = (json \ "name").asOpt[String].getOrElse("")
      val basisOperation =
        (json \ "basisOperation").asOpt[BasisOperation](basisOperationReads).getOrElse(OtherBasisOperation)
      val status = employmentStatus(json)
      val iyaCy = (json \ "inYearAdjustmentIntoCY").asOpt[BigDecimal].getOrElse(BigDecimal(0))
      val totalIya = (json \ "totalInYearAdjustment").asOpt[BigDecimal].getOrElse(BigDecimal(0))
      val iyaCyPlusOne = (json \ "inYearAdjustmentIntoCYPlusOne").asOpt[BigDecimal].getOrElse(BigDecimal(0))
      JsSuccess(
        TaxCodeIncome(
          incomeSourceType,
          employmentId,
          amount,
          description,
          taxCode,
          name,
          basisOperation,
          status,
          iyaCy,
          totalIya,
          iyaCyPlusOne))
    }
  }`

So the reading from the Hods has already been covered and we would be implementing/testing something that will likely never be used.

Reply:
I have to disagree with you on this point. The test will assert that reading and writing do not change data that should not be changed during the serialization process. For example if you were to change the writer to be
```
JsObject(
        List(
          "componentType"                 -> Json.toJson(o.componentType),
          "employmentId"                  -> Json.toJson(o.employmentId),
          "amount"                        -> Json.toJson(o.amount),
          "description"                   -> Json.toJson(o.description),
          "taxCode"                       -> Json.toJson(o.taxCodeWithEmergencySuffix),
          "name"                          -> Json.toJson(o.name),
          "basisOperation"                -> Json.toJson(o.basisOperation),
          "status"                        -> Json.toJson(o.status),
          "inYearAdjustmentIntoCY"        -> Json.toJson(o.inYearAdjustmentIntoCY),
          "totalInYearAdjustment"         -> Json.toJson(o.totalInYearAdjustment),
          "inYearAdjustmentIntoCYPlusOne" -> Json.toJson(o.totalInYearAdjustment), // this value is wrong
          "iabdUpdateSource"              -> Json.toJson(o.iabdUpdateSource),
          "updateNotificationDate"        -> Json.toJson(o.updateNotificationDate),
          "updateActionDate"              -> Json.toJson(o.updateActionDate)
        ).filter {
          case (_, JsNull) => false
          case _           => true
        }
      )
```
none of the existing tests would fail when clearly some should. However if you implement the test as I have shown above you would get a failing test. It would also help safe guard future changes to the serialization and model implementation